src/Http/Middleware/BodyParserMiddleware.php

```diff
@@ -63,7 +64,13 @@ class BodyParserMiddleware implements MiddlewareInterface
      */
     public function __construct(array $options = [])
     {
-        $options += ['json' => true, 'xml' => false, 'methods' => null];
+        $options += ['form-urlencoded' => true, 'json' => true, 'xml' => false, 'methods' => null];
+        if ($options['form-urlencoded']) {
```

Comment:
This change is going to trip people up as they upgrade. I think we can mitigate that a bit by updating the application skeleton now like you suggested. We could also add a deprecation warning in 4.0.x or 4.1.0 that ServerRequest parsing PUT/DELETE/PATCH data is going to move and that they should adopt the middleware.

Reply:
I have created a PR for the app skeleton. Adding deprecation warning is tricky. We tried doing that for `RequestHandler` where JSON parsing was removed but it caused too many false positives and we had to eventually get rid of it. It's difficult to distinguish whether a request body hasn't been parsed at all or an empty request body has been parsed.

Reply:
Yeah. I'm concerned we'll break applications in a way that is hard for application developers to detect.

Reply:
In that case I can move this form-urlencoded parsing too into `ServerRequestFactory` along with other things.

Reply:
That would be great. Having two copies isn't great but it gives us a low-risk path forward.

Reply:
> Having two copies isn't great

It isn't, so if we are moving it to `ServerRequestFactory` then I wouldn't update `BodyParserMiddleware`, otherwise the input would be unnecessarily parsed twice.

Reply:
Ok. Let's keep form-encoded in the request factory as it is a shim around a php limitation.